### PR TITLE
Add SelectObjectContentAsync API call

### LINKED ISSUE #359: Select object content
Hi,
do you plan to add select object content to the .NET API as well? Is there already any timeline for that? Thank you for the info

### PR DESCRIPTION
**Pull request by poornas**
fixes #359

### CODE REVIEW

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                var n =  ReadFromStream(prelude);
+                numBytesRead += n;
+                n =  ReadFromStream(preludeCRC);
+                numBytesRead += n;
+                var inputArray = new byte[prelude.Length + preludeCRC.Length];
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+
+                // write real data to inputArray
+                Crc32Algorithm.ComputeAndWriteToEnd(inputArray); // last 4 bytes contains CRC
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: @poornas We are actually overwriting the last 4 bytes (preludeCRC) with the calculated CRC in this function `ComputeAndWriteToEnd`

```
        /// <summary>
        /// Computes CRC-32 from input buffer and writes it after end of data (buffer should have 4 bytes reserved space for it). Can be used in conjunction with <see cref="IsValidWithCrcAtEnd(byte[],int,int)"/>
        /// </summary>
        /// <param name="input">Input buffer with data to be checksummed.</param>
        /// <param name="offset">Offset of the input data within the buffer.</param>
        /// <param name="length">Length of the input data in the buffer.</param>
        /// <returns>CRC-32 of the data in the buffer.</returns>
        public static uint ComputeAndWriteToEnd(byte[] input, int offset, int length)
        {
            if (length + 4 > input.Length)
                throw new ArgumentOutOfRangeException("length", "Length of data should be less than array length - 4 bytes of CRC data");
            var crc = Append(0, input, offset, length);
            var r = offset + length;
            input[r] = (byte)crc;
            input[r + 1] = (byte)(crc >> 8);
            input[r + 2] = (byte)(crc >> 16);
            input[r + 3] = (byte)(crc >> 24);
            return crc;
        }
 
        /// <summary>
        /// Computes CRC-32 from input buffer - 4 bytes and writes it as last 4 bytes of buffer. Can be used in conjunction with <see cref="IsValidWithCrcAtEnd(byte[])"/>
        /// </summary>
        /// <param name="input">Input buffer with data to be checksummed.</param>
        /// <returns>CRC-32 of the data in the buffer.</returns>
        public static uint ComputeAndWriteToEnd(byte[] input)
        {
            if (input.Length < 4)
                throw new ArgumentOutOfRangeException("input", "input array should be 4 bytes at least");
            return ComputeAndWriteToEnd(input, 0, input.Length - 4);
        }
```
Our `inputArray` here which is of 12bytes length has 8bytes(prelude)+4(preludeCRC) and this function `ComputeAndWriteToEnd` will calculate the CRC for (n - tailing 4) bytes and write the CRC to the last 4 bytes in inputArray. Which will overwrite the preludeCRC.

I think we should not copy the preludeCRC before validating. ie, We can do something like

```
var inputArray = new byte[prelude.Length + 4]; // 4 is the space for storing the calculated CRC
System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length); // Copy the prelude only
Crc32Algorithm.ComputeAndWriteToEnd(inputArray); 
// This will calculate the CRC for the prelude (first 8 bytes of the inputArray) and write the
// calculated CRC in the last 4 bytes of the inputArray
..
..
// We can then check if the last 4 bytes match the `preludeCRC` obtained from the input stream.
// (OR)
// We could calculate the crc by `var crc = Append(0, input, offset, length);` and then check if it 
// matches preludeCRC from the input stream.
```

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 2]
```diff
@@ -182,6 +182,41 @@ public async Task GetObjectAsync(string bucketName, string objectName, string fi
...
+            var selectReqBytes = System.Text.Encoding.UTF8.GetBytes(opts.MarshalXML());
+
+            string resourcePath = "?select=&select-type=2";
+            var request = await this.CreateRequest(Method.POST, bucketName,
+                                                    objectName: objectName,
+                                                    resourcePath: resourcePath,
+                                                    headerMap: sseHeaders)
+                                    .ConfigureAwait(false);
+
+            request.AddParameter("application/xml", selectReqBytes, ParameterType.RequestBody);
```
[the remark this answers is not in the document]
Comment (poornas) [reply Minio/ApiEndpoints/ObjectOperations.cs]: [here](https://github.com/minio/minio-dotnet/blob/master/Minio/V4Authenticator.cs#L112) - before authentication.

File: Minio/DataModel/Select/SelectResponseStream.cs (hunk 1 above)
Comment (poornas) [reply Minio/DataModel/Select/SelectResponseStream.cs]: good catch - fixed this

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                var n =  ReadFromStream(prelude);
+                numBytesRead += n;
+                n =  ReadFromStream(preludeCRC);
+                numBytesRead += n;
+                var inputArray = new byte[prelude.Length + preludeCRC.Length + 4];
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+
+                // write real data to inputArray
+                Crc32Algorithm.ComputeAndWriteToEnd(inputArray); // last 4 bytes contains CRC
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: Here the `computeAndWriteToEnd` calculates the CRC for the first 8(prelude) + 4(preludeCRC) bytes and writes the CRC in the last 4 byte block. We should not included the preludeCRC in the CRC calculation.

let the inputArray be `var inputArray = new byte[prelude.Length + 4];` and let us copy the prelude alone to the inputArray and let `ComputeAndWriteToEnd` calculate the CRC for the first 8bytes (prelude) and write that in the last 4 byte block. We could then verify this calculated CRC in the last 4bytes with the preludeCRC that we got from the stream - (`n =  ReadFromStream(preludeCRC);`)

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+                numBytesRead += num;
+                num = ReadFromStream(messageCRC);
+                // now verify message CRC
+                inputArray = new byte[totalLength + 4];
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+                System.Buffer.BlockCopy(headers,0,inputArray,prelude.Length+ preludeCRC.Length, headerLength);
+                System.Buffer.BlockCopy(payload,0,inputArray,prelude.Length+ preludeCRC.Length+headerLength, payloadLength);
+                System.Buffer.BlockCopy(messageCRC,0,inputArray, totalLength - 4,messageCRC.Length);
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: We should exclude messageCRC for the crc calculation here as well

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+                    {
+                        Console.WriteLine("payload|"+Encoding.UTF7.GetString(payload));
+                        StatsMessage stats = new StatsMessage();
+                        using (var stream = new MemoryStream(payload))
+                        stats = (StatsMessage)new XmlSerializer(typeof(StatsMessage)).Deserialize(stream);
+                        this.Stats = stats;
+                    }
+                    if (value.Equals("Records"))
+                    {
+                        this.Payload.Write(payload,0, payloadLength);
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: If i understand correctly, `Payload` is the response writer that the caller/user will be listening to? @poornas

File: Minio/V4Authenticator.cs [hunk 6]
```diff
@@ -383,22 +384,22 @@ internal string GetPresignCanonicalRequest(IRestClient client, IRestRequest requ
             // QUERY
             if (path.Length == 2)
             {
-                var parameterString = path[1];
-                var parameterList = parameterString.Split('&');
-                var sortedQueries = new SortedSet<string>();
-                foreach (string individualParameterString in parameterList)
+                var queryParams = path[1].Split('&').Select(p => p.Split('='))
```
Comment (Praveenrajmani) [inline Minio/V4Authenticator.cs]: This change is not related to Select right?

File: Minio/DataModel/Select/SelectResponseStream.cs (hunk 3 above)
Comment (poornas) [reply Minio/DataModel/Select/SelectResponseStream.cs]: fixed.

File: Minio/DataModel/Select/SelectResponseStream.cs (hunk 5 above)
Comment (poornas) [reply Minio/DataModel/Select/SelectResponseStream.cs]: that's correct.

File: Minio/V4Authenticator.cs (hunk 6 above)
Comment (poornas) [reply Minio/V4Authenticator.cs]: Query parameters were not being canonicalized correctly - a preexisting bug exposed while implementing the select API. Hence this is a part of the PR.

File: Minio/DataModel/Select/SelectResponseStream.cs (hunk 4 above)
Comment (poornas) [reply Minio/DataModel/Select/SelectResponseStream.cs]: fixed

**Comment by poornas** [comment]
PR updated @Praveenrajmani & @krisis

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+                numBytesRead += num;
+                num = ReadFromStream(messageCRC);
+                var messageCRCBytes = messageCRC.ToArray();
+                if (BitConverter.IsLittleEndian)
+                {
+                    Array.Reverse(messageCRCBytes); 
+                }
+                // now verify message CRC
+                inputArray = new byte[totalLength];
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: Should we allocate 4 more bytes for the messageCRC check? like we did for the preludeCRC above?

File: Minio/DataModel/Select/SelectResponseStream.cs [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                System.Buffer.BlockCopy(prelude,0,inputArray,0, prelude.Length);
+                System.Buffer.BlockCopy(preludeCRC,0,inputArray,prelude.Length, preludeCRC.Length);
+                System.Buffer.BlockCopy(headers,0,inputArray,prelude.Length+ preludeCRC.Length, headerLength);
+                System.Buffer.BlockCopy(payload,0,inputArray,prelude.Length+ preludeCRC.Length+headerLength, payloadLength);
+
+                // write real data to inputArray
+                Crc32Algorithm.ComputeAndWriteToEnd(inputArray); // last 4 bytes contains CRC
```
Comment (Praveenrajmani) [inline Minio/DataModel/Select/SelectResponseStream.cs]: So, this overwrites the last four bytes of the payload. `inputArray = new byte[totalLength+4];` may fix this.

Comment (poornas) [reply Minio/DataModel/Select/SelectResponseStream.cs]: @Praveenrajmani , totalLength includes the entire 16 byte overhead including 4 for the messageCRC. The messageCRC is not copied into the inputArray, leaving 4 bytes for the CRC32 algorithm to store the computed value. Seems correct to me.
